docs/proposals/mapper-redesign.md

```diff
@@ -0,0 +1,139 @@
+The process above makes the routine complex.
+2. For one node, devices combined to it are almost stable, the change should be minor, so the device configmap is almost unchangeable. The value to deal with this change is minor.
+3. Furthermore, we could process the configmap change with K8S style method: restart the container which is easy and simple.
+
+### Architecture
+Components are as below:
+<img src="../images/mapper/components.PNG">
+
+* Configmap parser: parse the configmap
+* Driver: device visit (init, read, write, healthy status or connection status)
+* Event process: subscribe, publish Mqtt, message receive and process
+* Timer: call the function periodically
+
+***The configmap parser, event process, and timer could be extracted as SDK.***
+
+Other two parts are:
+* Flags: configurations including mqtt user/password/Certification
+* Makefile: unified compile all mappers or compile one mapper, it should be within the main makefile
+
+### Routine
+The main routine is as:
+<img src="../images/mapper/routine.PNG">
+
+The message interaction among event bus, mqtt broker, mapper, and 3-rd applications is as:
+<img src="../images/mapper/message.PNG">
+
+### Topics
+| Topic  | Publisher  | Subscriber  | Description  | Requirement  |
+| ------------ | ------------ | ------------ | ------------ | ------------ |
+|$hw/events/device/+/twin/update/delta   |edgecore   |mapper   |Notify that the expected value changed   |Yes   |
+|$hw/events/device/+/twin/update   |mapper   |edgecore   |Update the expected value to cloud  |Yes   |
+|$hw/events/device/+/twin/update/result   |edgecore   |mapper   |If edgecore receive the update result   |Optional   |
+|$ke/events/device/+/data/update   |mapper   |3-rd application   |Update data required by 3-rd   |Yes   |
+|$hw/events/device/+/state/update  |mapper   |edgecore   |Update device status   |Yes   |
+|$hw/events/node/+/membership/updated   |edgecore   |mapper   |Notify  device add/remove   |Optional   |
+|$hw/events/node/+/membership/get   |mapper   |edgecore   |Query device list   |Optional   |
+|$hw/events/node/+/membership/get/result   |edgecore   |mapper   |Result of querying device list   |Optional   |
+|SYS/dis/upload_records   |mapper/applications   |cloud   |Upload records to the cloud   |Reserved   |
+
+### Security
+Mqtt connection supports both user/password and certification mode.
+
+In the user/password mode, the user/password should be stored in a configuration file.
+
+In the certification mode, we will use one side authentication. The certification of mqtt client is issued by your own CA. The mqtt broker will validate the certification of mqtt client with the CA.
+
+### Miscellaneous
+#### Device status definition
+```go
+const DeviceStatus {
+    DEVSTOK
+    DEVSTERR        /*Expected value is not equal as setting*/
+    DEVSTDISCONN    
+    DEVSTUNHEALTHY  /*Unhealthy status from device*/
+}
```

Comment:
Not quite understand the code here, what's the value type of the consts?

Reply:
Just want to present all status that reported to the device controller. In the code, the type is uint.